mdtraj/core/topology.py

```diff
@@ -88,12 +88,17 @@ def _topology_from_subset(topology, atom_indices):
 
     for chain in topology._chains:
         newChain = newTopology.add_chain()
+        previous_residue = None
         for residue in chain._residues:
             resSeq = getattr(residue, 'resSeq', None) or residue.index
-            newResidue = newTopology.add_residue(residue.name, newChain,
-                                                 resSeq)
             for atom in residue._atoms:
                 if atom.index in atom_indices:
+                    if not residue == previous_residue:
+                        newResidue = newTopology.add_residue(residue.name, newChain,
+                                                             resSeq)
+                        newResidue.segment_id = residue.segment_id
```

Comment:
I believe the reason for moving the new residue command is related to what happens when there are no selected atoms from a residue. The add_residue command increments the number of residues and the residue's serial number, and then decides whether it's empty or not. It seems like most of the time you want your new topology to have the number of residues equal to the number of non-empty residues. The resSeq takes care of the fact that moving this inside the loop renumbers the residues, since if you select by resSeq it should still be the old numbering scheme. Does that seem right?